substrate/frame/staking/src/slashing.rs

```diff
@@ -321,17 +321,48 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
 }
 
 /// Inform the [`DisablingStrategy`] implementation about the new offender and disable the list of
-/// validators provided by [`make_disabling_decision`].
+/// validators provided by [`decision`].
 fn add_offending_validator<T: Config>(params: &SlashParams<T>) {
 	DisabledValidators::<T>::mutate(|disabled| {
-		if let Some(offender) =
-			T::DisablingStrategy::decision(params.stash, params.slash_era, &disabled)
-		{
-			// Add the validator to `DisabledValidators` and disable it. Do nothing if it is
-			// already disabled.
-			if let Err(index) = disabled.binary_search_by_key(&offender, |index| *index) {
-				disabled.insert(index, offender);
-				T::SessionInterface::disable_validator(offender);
+		let decision =
+			T::DisablingStrategy::decision(params.stash, params.slash, params.slash_era, &disabled);
+
+		if let Some(offender_idx) = decision.disable {
+			// Check if the offender is already disabled
+			match disabled.binary_search_by_key(&offender_idx, |(index, _)| *index) {
+				// Offender is already disabled, update severity if the new one is higher
+				Ok(index) => {
+					let (_, old_severity) = &mut disabled[index];
+					if params.slash > *old_severity {
```

Comment:
We are indeed only depositing slashes that are the highest in an era. That is a good catch. The [`UpToLimitWithReEnablingDisablingStrategy`] simply considers the most recent given to us slash if it was larger then previous ones seen. So re-enabling has a baked in max assumption.

This means that it would still work on cumulative slashes but possibly not as one intends (AKA only use the highest slash instead of total slashes).

I added this assumption explicitly mentioned in the [`UpToLimitWithReEnablingDisablingStrategy`]

```rust
/// This strategy is not based on cumulative severity of offences but only on the severity of the highest offence.
/// Offender first committing a 25% offence and then a 50% offence will be treated the same as an offender committing 50%
/// offence.
```

If someone would like to have a cumulative disabling version this would require a new disablingStrategy impl.